lib/spack/spack/relocate_text.py

```diff
@@ -0,0 +1,288 @@
+# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+"""This module contains pure-Python classes and functions for replacing
+paths inside text files and binaries."""
+
+import re
+from collections import OrderedDict
+from typing import Dict, Union
+
+import spack.error
+
+Prefix = Union[str, bytes]
+
+
+def encode_path(p: Prefix) -> bytes:
+    return p if isinstance(p, bytes) else p.encode("utf-8")
+
+
+def _prefix_to_prefix_as_bytes(prefix_to_prefix) -> Dict[bytes, bytes]:
+    return OrderedDict((encode_path(k), encode_path(v)) for (k, v) in prefix_to_prefix.items())
+
+
+def utf8_path_to_binary_regex(prefix: str):
+    """Create a binary regex that matches the input path in utf8"""
+    prefix_bytes = re.escape(prefix).encode("utf-8")
+    return re.compile(b"(?<![\\w\\-_/])([\\w\\-_]*?)%s([\\w\\-_/]*)" % prefix_bytes)
+
+
+def _byte_strings_to_single_binary_regex(prefixes):
+    all_prefixes = b"|".join(re.escape(p) for p in prefixes)
+    return re.compile(b"(?<![\\w\\-_/])([\\w\\-_]*?)(%s)([\\w\\-_/]*)" % all_prefixes)
+
+
+def utf8_paths_to_single_binary_regex(prefixes):
+    """Create a (binary) regex that matches any input path in utf8"""
+    return _byte_strings_to_single_binary_regex(p.encode("utf-8") for p in prefixes)
+
+
+def filter_identity_mappings(prefix_to_prefix):
+    """Drop mappings that are not changed."""
+    # NOTE: we don't guard against the following case:
+    # [/abc/def -> /abc/def, /abc -> /x] *will* be simplified to
+    # [/abc -> /x], meaning that after this simplification /abc/def will be
+    # mapped to /x/def instead of /abc/def. This should not be a problem.
+    return OrderedDict((k, v) for (k, v) in prefix_to_prefix.items() if k != v)
+
+
+class PrefixReplacer:
+    """Base class for applying a prefix to prefix map
+    to a list of binaries or text files.
+    Child classes implement _apply_to_file to do the
+    actual work, which is different when it comes to
+    binaries and text files."""
+
+    def __init__(self, prefix_to_prefix: Dict[bytes, bytes]):
+        """
+        Arguments:
+
+            prefix_to_prefix (OrderedDict):
+
+                A ordered mapping from prefix to prefix. The order is
+                relevant to support substring fallbacks, for example
+                [("/first/sub", "/x"), ("/first", "/y")] will ensure
+                /first/sub is matched and replaced before /first.
+        """
+        self.prefix_to_prefix = filter_identity_mappings(prefix_to_prefix)
+
+    @property
+    def is_noop(self) -> bool:
+        """Returns true when the prefix to prefix map
+        is mapping everything to the same location (identity)
+        or there are no prefixes to replace."""
+        return not bool(self.prefix_to_prefix)
```

Comment:
Another very minor comment. I don't think it's necessary to cast to bool here. Using:
```suggestion
        return not self.prefix_to_prefix
```
should have the same effect.